### PR TITLE
Stack nested functions in jitted loop bodies

### PR DESCRIPTION
**Pull request by pleath**
Allow jitted loop bodies to make use of stack nested functions if indicated by the byte code. The prologue code will initialize a pointer to the interpreter instance's stack nested functions, and each NewScFunc that allocates on the stack will make use of the pointer. The IRBuilder inserts an instruction that refers to the closure environment and the stack nested function pointer and feeds the NewScFunc instruction, since NewScFunc already takes 2 source operands in some cases.

### CODE REVIEW

File: lib/Backend/Func.cpp [hunk 1]
```diff
@@ -147,25 +147,18 @@ Func::Func(JitArenaAllocator *alloc, CodeGenWorkItem* workItem, const Js::Functi
...
-            stackNestedFunc = true;
-            this->GetTopFunc()->hasAnyStackNestedFunc = true;
-        }
+        // doBackendArgumentsOptimization bit is set when there is no eval inside a function
+        // as determined by the bytecode generator.
+        SetHasStackArgs(true);
     }
-    else
+    if (doStackNestedFunc && m_jnFunction->GetNestedCount() != 0 &&
+        (this->IsTopFunc() || this->GetTopFunc()->m_workItem->Type() != JsLoopBodyWorkItemType)) // make sure none of the functions inlined in a jitted loop body allocate nested functions on the stack
```
Comment (rajatd) [inline lib/Backend/Func.cpp]: >this->GetTopFunc()->m_workItem->Type() != JsLoopBodyWorkItemType [](start = 30, length = 64)

Can we now get rid of this condition for inlinees in a jitted loop body?

**Comment by rajatd** [comment]
:shipit:

File: lib/Backend/Func.cpp (hunk 1 above)
Comment (pleath) [inline lib/Backend/Func.cpp]: I don't think so. An inlinee can't make use of the functions allocated by the interpreter instance, of course. Are we allocating nested functions on the parent's frame if the top-level function is a jitted loop body? (If we're not, we could, but I didn't want to do that as part of this change.)

Comment (rajatd) [inline lib/Backend/Func.cpp]: Right now, we are not allocating nested functions in an inlinee on the stack at all if the inlinee was in a jitted loop body. I wanted to know if we could do that now. Agreed that it is out of scope of this change.

---
In reply to: [77870462](https://github.com/Microsoft/ChakraCore/pull/1547#discussion_r77870462) [](ancestors = 77870462)
